### PR TITLE
solution hw-spring-05

### PR DESCRIPTION
**Pull request by RomanHreits**

### CODE REVIEW

File: src/main/java/com/cinema/controller/AuthenticationController.java [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+
+import com.cinema.model.dto.UserAuthenticationDto;
+import com.cinema.security.AuthenticationService;
+import org.springframework.web.bind.annotation.PostMapping;
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RestController;
+
+@RestController
+public class AuthenticationController {
+    private final AuthenticationService authentication;
```
Comment (Rommelua) [inline src/main/java/com/cinema/controller/AuthenticationController.java]: better change name to "service" or "authenticationService"

File: src/main/java/com/cinema/controller/MovieSessionController.java [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+    @PostMapping
+    public void add(@RequestBody MovieSessionRequestDto requestDto) {
+        MovieSession movieSession = mapper.mapFromReqDtoToEntity(requestDto);
+        movieSessionService.add(movieSession);
+    }
+
+    @GetMapping("/available")
+    public List<MovieSessionResponseDto> getByParams(@RequestParam Long movieId,
+                                                     @RequestParam String date) {
```
Comment (Rommelua) [inline src/main/java/com/cinema/controller/MovieSessionController.java]: You can use LocalDate as method parametr. String is parsing to LocalDate on fly.

File: src/main/java/com/cinema/mapper/GenericMapper.java [hunk 3]
```diff
@@ -0,0 +1,9 @@
+package com.cinema.mapper;
+
+public interface GenericMapper<T, K, V> {
+    T mapFromReqDtoToEntity(K dto);
+
+    K mapToReqDto(T entity);
```
Comment (Rommelua) [inline src/main/java/com/cinema/mapper/GenericMapper.java]: Why do you need map entity to ReqDto? You receive it from frontend and ReqDto have to be mapped to entity, not vice verse

File: src/main/java/com/cinema/mapper/GenericMapper.java [hunk 4]
```diff
@@ -0,0 +1,9 @@
+package com.cinema.mapper;
+
+public interface GenericMapper<T, K, V> {
```
Comment (Rommelua) [inline src/main/java/com/cinema/mapper/GenericMapper.java]: To my opinion this interface and its children are redundant, classes would be enough

File: src/main/java/com/cinema/model/dto/MovieDtoRequest.java [hunk 5]
```diff
@@ -0,0 +1,31 @@
+package com.cinema.model.dto;
+
+public class MovieDtoRequest {
+    private Long id;
```
Comment (Rommelua) [inline src/main/java/com/cinema/model/dto/MovieDtoRequest.java]: We don't need id in request DTO - it is new object and id will be created by DB

File: src/main/java/com/cinema/controller/AuthenticationController.java [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+
+import com.cinema.model.dto.UserAuthenticationDto;
+import com.cinema.security.AuthenticationService;
+import org.springframework.web.bind.annotation.PostMapping;
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RestController;
+
+@RestController
+public class AuthenticationController {
+    private final AuthenticationService service;
```
Comment (boroda4436) [inline src/main/java/com/cinema/controller/AuthenticationController.java]:
```suggestion
    private final AuthenticationService authService;
```

File: src/main/java/com/cinema/controller/AuthenticationController.java [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+
+    public AuthenticationController(AuthenticationService service) {
+        this.service = service;
+    }
+
+    @PostMapping("/register")
+    public void register(@RequestBody UserAuthenticationDto userAuthenticationDto) {
```
Comment (boroda4436) [inline src/main/java/com/cinema/controller/AuthenticationController.java]: UserAuthenticationDto and UserRegistrationDto could be different. For example, in the `UserRegistrationDto` you could have fields  `login`, `password`, and `repeatPassword`. But in `UserAuthenticationDto` you could have only `login` and `password`.

File: src/main/java/com/cinema/controller/MovieController.java [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+    private final MovieService movieService;
+    private final ModelMapper modelMapper;
+
+    public MovieController(MovieService movieService, ModelMapper modelMapper) {
+        this.movieService = movieService;
+        this.modelMapper = modelMapper;
+    }
+
+    @PostMapping
+    public void create(@RequestBody MovieDtoRequest movieDto) {
```
Comment (boroda4436) [inline src/main/java/com/cinema/controller/MovieController.java]:
```suggestion
    public void create(@RequestBody MovieRequestDto movieDto) {
```

File: src/main/java/com/cinema/controller/MovieController.java [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+        Movie movie = modelMapper.map(movieDto, Movie.class);
+        movieService.add(movie);
+    }
+
+    @GetMapping
+    public List<MovieDtoResponse> getAll() {
```
Comment (boroda4436) [inline src/main/java/com/cinema/controller/MovieController.java]:
```suggestion
    public List<MovieResponseDto> getAll() {
```

File: src/main/java/com/cinema/controller/MovieSessionController.java (hunk 2 above)
Comment (boroda4436) [reply src/main/java/com/cinema/controller/MovieSessionController.java]: agree

File: src/main/java/com/cinema/service/MovieService.java [hunk 10]
```diff
@@ -7,4 +7,6 @@
     Movie add(Movie movie);
     
     List<Movie> getAll();
+
+    Movie getById(Long id);
```
Comment (boroda4436) [inline src/main/java/com/cinema/service/MovieService.java]:
```suggestion
    Movie get(Long id);
```

File: src/main/java/com/cinema/service/CinemaHallService.java [hunk 11]
```diff
@@ -7,4 +7,6 @@
     CinemaHall add(CinemaHall cinemaHall);
     
     List<CinemaHall> getAll();
+
+    CinemaHall getById(Long id);
```
Comment (boroda4436) [inline src/main/java/com/cinema/service/CinemaHallService.java]:
```suggestion
    CinemaHall get(Long id);
```

File: src/main/java/com/cinema/service/impl/CinemaHallServiceImpl.java [hunk 12]
```diff
@@ -23,4 +23,9 @@ public CinemaHall add(CinemaHall cinemaHall) {
     public List<CinemaHall> getAll() {
         return cinemaHallDao.getAll();
     }
+
+    @Override
+    public CinemaHall getById(Long id) {
```
Comment (boroda4436) [inline src/main/java/com/cinema/service/impl/CinemaHallServiceImpl.java]:
```suggestion
    public CinemaHall get(Long id) {
```

File: src/main/java/com/cinema/service/impl/MovieServiceImpl.java [hunk 13]
```diff
@@ -23,4 +23,9 @@ public Movie add(Movie movie) {
     public List<Movie> getAll() {
         return movieDao.getAll();
     }
+
+    @Override
+    public Movie getById(Long id) {
```
Comment (boroda4436) [inline src/main/java/com/cinema/service/impl/MovieServiceImpl.java]: the same

**Review by boroda4436 — CHANGES_REQUESTED** [review]

File: src/main/java/com/cinema/dao/CinemaHallDao.java [hunk 14]
```diff
@@ -7,4 +7,6 @@
     CinemaHall add(CinemaHall cinemaHall);
 
     List<CinemaHall> getAll();
+
+    CinemaHall getById(Long id);
```
Comment (Sofasmile) [inline src/main/java/com/cinema/dao/CinemaHallDao.java]: ну вже якщо ти змінив getById в одній частині то варто змінити скрізь))

File: src/main/java/com/cinema/model/dto/MovieRequestDto.java [hunk 15]
```diff
@@ -0,0 +1,23 @@
+package com.cinema.model.dto;
+
+public class MovieRequestDto {
+
```
Comment (Sofasmile) [inline src/main/java/com/cinema/model/dto/MovieRequestDto.java]: remove empty line

File: src/main/java/com/cinema/model/dto/MovieSessionRequestDto.java [hunk 16]
```diff
@@ -0,0 +1,32 @@
+package com.cinema.model.dto;
+
+public class MovieSessionRequestDto {
+
```
Comment (Sofasmile) [inline src/main/java/com/cinema/model/dto/MovieSessionRequestDto.java]: same

File: src/main/java/com/cinema/model/dto/MovieSessionResponseDto.java [hunk 17]
```diff
@@ -0,0 +1,40 @@
+package com.cinema.model.dto;
+
+public class MovieSessionResponseDto {
+    private Long movieSessionId;
+    private String movieTitle;
```
Comment (Sofasmile) [inline src/main/java/com/cinema/model/dto/MovieSessionResponseDto.java:6]: let's add movie id too

File: src/main/java/com/cinema/model/dto/UserRegistrationDto.java [hunk 18]
```diff
@@ -0,0 +1,23 @@
+package com.cinema.model.dto;
+
+public class UserRegistrationDto {
+
```
Comment (Sofasmile) [inline src/main/java/com/cinema/model/dto/UserRegistrationDto.java]: and here

File: src/main/java/com/cinema/mapper/GenericMapper.java (hunk 4 above)
Comment (Sofasmile) [reply src/main/java/com/cinema/mapper/GenericMapper.java]: I agree with @Rommelua GenericMapper interface is redundant
we can use only classes

File: src/main/java/com/cinema/dao/CinemaHallDao.java (hunk 14 above)
Comment (RomanHreits) [reply src/main/java/com/cinema/dao/CinemaHallDao.java]: неуважний(

**Review by Sofasmile — APPROVED** [review]

**Pull request merged**